### PR TITLE
Improve the README.md particularly around the signing/encryption features

### PR DESCRIPTION
**Pull request by johnnyshields**
Also adds a missed line to UPGRADING.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -522,72 +525,169 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
 ```
 
-The settings related to sign are stored in the `security` attribute of the settings:
+## Signing and Encryption
+
+Ruby SAML supports supports the signing and encryption functionality:
+
+1. Signing your SP Metadata XML
+2. Signing your SP SAML messages
+3. Encrypting IdP Assertion messages, and decrypting them upon receipt (EncryptedAssertion)
```
Comment (pitbulk) [inline README.md:565]: Toolkit does not support  "Encrypting IdP Assertion messages", only decrypt them

File: README.md [hunk 2]
```diff
@@ -522,72 +525,169 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
+4. Verifying signatures on IdP Assertion messages
```
Comment (pitbulk) [inline README.md:566]: Verifying SAML Message signatures or SAML Assertions generated by the IdP

File: README.md [hunk 3]
```diff
@@ -522,72 +525,169 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
-Signatures will be handled for both `HTTP-Redirect` and `HTTP-Redirect` Bindings.
+Signatures will be handled automatically for both `HTTP-Redirect` and `HTTP-Redirect` Binding.
 Note that the RelayState parameter is used when creating the Signature on the `HTTP-Redirect` Binding.
 Remember to provide it to the Signature builder if you are sending a `GET RelayState` parameter or the
 signature validation process will fail at the Identity Provider.
 
-The Service Provider will sign the request/responses with its private key.
-The Identity Provider will validate the sign of the received request/responses with the public x509 cert of the
-Service Provider.
+#### Encrypting SAML Assertions
```
Comment (pitbulk) [inline README.md:623]: Decrypting SAML Assertions

File: README.md [hunk 4]
```diff
@@ -522,72 +525,169 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
+public cert of the Service Provider. The Service Provider will decrypt the EncryptedAssertion with its private key.
+
+You may enable EncryptedAssertion as follows. This will add `<md:KeyDescriptor use="encrytion">` to your
+SP Metadata XML, to be read by the IdP.
+
+```ruby
+  settings.certificate = "CERTIFICATE TEXT WITH BEGIN/END HEADER AND FOOTER"
+  settings.private_key = "PRIVATE KEY TEXT WITH BEGIN/END HEADER AND FOOTER"
+
+  settings.security[:want_assertions_encrypted] = true # Enable EncryptedAssertion
```
Comment (pitbulk) [inline README.md:635]: This parameter is to invalidate any SAML Message with no encrypted assertion, so it is not directly related with the decrypt process.

File: README.md [hunk 5]
```diff
@@ -522,72 +525,169 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
-  settings.security[:authn_requests_signed]   = true     # Enable or not signature on AuthNRequest
-  settings.security[:logout_requests_signed]  = true     # Enable or not signature on Logout Request
-  settings.security[:logout_responses_signed] = true     # Enable or not signature on Logout Response
-  settings.security[:want_assertions_signed]  = true     # Enable or not the requirement of signed assertion
-  settings.security[:metadata_signed]         = true     # Enable or not signature on Metadata
+  settings.certificate = "CERTIFICATE TEXT WITH BEGIN/END HEADER AND FOOTER"
+  settings.private_key = "PRIVATE KEY TEXT WITH BEGIN/END HEADER AND FOOTER"
+```
+
+Note that the same certificate and private key are used for all SP encryption and signing-related functions.
```
Comment (pitbulk) [inline README.md:575]: Note that the certificate/private key that gonna be involved on the sign/validate encrypt/decrypt process gonna be the same.
IdP will encrypt using SP certificate and then SP will decrypt using the SP private_key.
SP will sign using the SP private_key, and the IdP will validate the signature with the SP certificate.
So you see that the SP certificate must be registered on the IdP side, to be used to encrypt and validate signatures.

File: README.md (hunk 2 above)
Comment (johnnyshields) [reply README.md:566]: Verifying signatures on SAML messages and IdP Assertions

File: README.md (hunk 5 above)
Comment (johnnyshields) [reply README.md:575]: I've edited it to:

> Note that the same certificate (and its associated private key) are used to perform
> all decryption and signing-related functions (1-4) above. Ruby SAML does not currently allow
> to specify different certificates for each function.

The key point here is the last sentence; we are simply saying that you can't decrypt with Cert A and sign with Cert B.

**Comment by johnnyshields** [comment]
All review feedback is addressed.

File: README.md [hunk 6]
```diff
@@ -522,72 +525,170 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
-  settings.private_key = "PRIVATE KEY TEXT WITH HEAD AND FOOT"
+  # Valid until => 2 days from now
+  # Cache duration = 604800s = 1 week
+  valid_until = Time.now + 172800
+  cache_duration = 604800
+  meta.generate(settings, false, valid_until, cache_duration)
 ```
 
-The settings related to sign are stored in the `security` attribute of the settings:
+## Signing and Encryption
```
Comment (pitbulk) [inline README.md:559]: ## Signing and Decryption

File: README.md [hunk 7]
```diff
@@ -522,72 +525,170 @@ To add a `saml:AuthnContextDeclRef`, define `settings.authn_context_decl_ref`.
...
+
+Ruby SAML supports supports the signing and encryption functionality:
```
Comment (pitbulk) [inline README.md:561]: Ruby SAML supports _supports_

**Review by pitbulk — APPROVED** [review]

**Pull request merged**